ruby: Map to_h fixes

skippy opened the pull request:
map needed a #to_h method

Also:
- moving back to strings for hash keys (a change from the previous commit #338) to keep it consistent with type definitions in maps, such as defining 'string' as the key type.  
- the behavior of RepeatedFields#each_index was not correct.  The array and enumerable
  behaviors are different (who knew?!), so making this equivalent to the array version
- merge in upstream changes from master.  **WARNING** under jruby the expected protobuf binary test fails, meaning it generates a different protobuf than the exact same msg run under MRi.  This seems to be an issue upstream in the java library than in this code base.
- Fixing jruby errors when run under jruby-9000.pre1.  Their is a jruby AST parse error that required odd code to work around in a few of the tests.
- Found a bug with the `encode_json` method.  When run under MRI, it does not return any default values, while JRuby does.  It should probably return default values.

skippy commented on ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java:
```diff
@@ -189,8 +190,25 @@ public IRubyObject index(ThreadContext context, IRubyObject key) {
+                return context.runtime.getTrue();
+            }
+
+            RubyHash other = (RubyHash)_other;
```
the reason we have another step here is that `#toHash` will convert all complex objects, such as `Map`, `Message`, and `RepeatedField` into hashes and arrays.  I believe this is what one would expect when calling `#to_h`. However, that means that if the `==` is comparing against something like what is in `ruby/tests/basic.rb#ln616` will fail unless we do a deeper check.

let me know what you think and if this is too complex and/or the assumptions on `#to_h` are wrong

cfallin replied:
Hmm -- I assume you actually hit the test failure then wrote this to fix it? I was under the impression that a Ruby hash would do a deep equality comparison already (using the `:==` method on values as appropriate) -- the [docs](http://ruby-doc.org/core-2.2.0/Hash.html#method-i-3D-3D) seem to say this too. Perhaps JRuby behaves differently?

skippy replied:
I thought we did need the deep comparison, to deal with cases where the underlying hash may be expanded out or not (and thus is the original map object).  Let me double check....

skippy replied:
hey @cfallin 

I did not first write a failing test case and then fix it.  The failure was already there @ `ruby/tests/basic.rb#ln616`

that assert fails without the change to `==`.  If you take out this code, you'll see the test case fail; it passes with this deeper logic.  

the root issue is the modifications I made to `#toHash`, which will convert complex child objects (Map/Message) into hashes rather than return the complex object.

I think this 'new' behavior is the expected behavior, but of course that is open to debate :)

skippy replied:
@cfallin another thought; the original logic would allow equivalency with a hash.  I'm not sure that is really the correct behavior.  So if we change it to where if the incoming obj is not equivalent to this, return false.

skippy commented on ruby/tests/repeated_field_test.rb:
```diff
@@ -570,11 +578,29 @@ def check_self_modifying_method(repeated_field, ref_array)
+  #  commented examples that fail in jruby are below
   def repeated_field_names(klass)
     klass.descriptor.find_all{|f| f.label == :repeated}.map(&:name)
+    # jruby failure cases
```
ugh

skippy commented on ruby/tests/repeated_field_test.rb:
```diff
@@ -2,19 +2,20 @@
 
 require 'google/protobuf'
 require 'test/unit'
+require_relative './support/message_definitions'
```
moved commonly used definitions over into a single file.

cfallin commented on ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java:
```diff
@@ -340,7 +358,20 @@ public IRubyObject dup(ThreadContext context) {
     @JRubyMethod(name = {"to_h", "to_hash"})
     public RubyHash toHash(ThreadContext context) {
-        return RubyHash.newHash(context.runtime, table, context.runtime.getNil());
+        // return RubyHash.newHash(context.runtime, table, context.runtime.getNil());
```
Remove commented-out line here

cfallin commented on ruby/src/main/java/com/google/protobuf/jruby/RubyMap.java:
```diff
@@ -189,8 +190,25 @@ public IRubyObject index(ThreadContext context, IRubyObject key) {
+            RubyHash other = (RubyHash)_other;
+            // we have complex objects, such as a message
+            // lets check to see if they are equivalent
+            for (IRubyObject key : table.keySet()) {
```
If we do need this deep comparison -- I think it's missing a check for the case where `other` has additional keys. E.g., say that `this` is `{}`, and `other` is `{ 'a' => MyMessage.new(...)}` -- the above `op_equal` returns false, so we fall through to here, the for-loop never iterates (`table` is empty), and we return `true`.

skippy replied:
fixed!

cfallin commented:
Hi @skippy, thanks for this! I agree that the fix to semantics makes sense. A few minor concerns above with implementation but overall direction is good.

skippy commented on ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java:
```diff
@@ -345,13 +345,18 @@ public IRubyObject toHash(ThreadContext context) {
-                if (value.respondsTo("to_h")) {
+                // Order here matters!  Enumerable fields, such as RepeatedField,
+                // respond to both #to_h and #to_a.  in that case, we want the #to_a
+                // behavior
```
this is an issue under jruby-9000, but not MRI ruby or earlier versions of jruby

skippy commented on ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java:
```diff
@@ -695,7 +700,7 @@ private String layoutInspect() {
             sb.append(Utils.unescapeIdentifier(fdef.getName()));
             sb.append(": ");
-            sb.append(getField(context, fdef).inspect());
+            sb.append(Helpers.invoke(context, getField(context, fdef), "inspect"));
```
this is a jruby-9000 fix.

skippy commented on ruby/tests/basic.rb:
```diff
@@ -476,75 +476,6 @@ def test_map_ctor
       assert m == {"a" => 1, "c" => 3, "b" => 2}
     end
 
-    def test_map_keytypes
```
started a `map_test.rb` file; it isn't complete, but I started to flesh it out

skippy commented on ruby/tests/message_test.rb:
```diff
@@ -0,0 +1,214 @@
+    if RUBY_PLATFORM == "java"
+      # FIXME: java has the correct result; the underlying C++ encode_json
+      # seems to be broken to some degree, where jruby returns the default values
+      # but the C library does not
```
**BUG** ?   @cfallin shall I go ahead and open a ticket?

skippy commented on ruby/tests/message_test.rb:
```diff
@@ -0,0 +1,214 @@
+
+    result = m.to_proto
+    expected_result = Base64.decode64("QgNmb29wCoABAYABAJoBA2JhcsIBBwoBYRICCAE=")
+    assert_equal expected_result, result
```
**BUG** mri and jruby return different values.  I'm not sure which one is right.  The Base64 encoded value here is from the MRI version.  @cfallin thoughts?

skippy commented:
@cfallin thanks for your support on this ticket.  let me know what you think.  Also be aware that I found what looks to be 2 bugs that are not related to this ticket.

cfallin commented:
@skippy Thanks for all the updates on this! Since the last back-and-forth, though, I've actually handed off Ruby duties to @haberman -- I am sure he'll be happy to take a look. Thanks!

haberman commented:
@skippy Sorry for the delay on this, I hadn't even realized this was still open. Could you bring me up to date on the overall status of this PR?

haberman commented:
@skippy Also could you possibly take a look at this test failure for an open PR and see if it related to what you are changing in this CL? I ask because it appears to be a failure related to `to_h`.

https://travis-ci.org/google/protobuf/jobs/66956115

That is a failure on this PR: https://github.com/google/protobuf/pull/500

Which is intended to fix this bug: https://github.com/google/protobuf/issues/493

skippy commented:
hey sir; I don't think this PR is causing the `RubyMap#hash` method to fail.  This PR changes the `#toHash` (as well as `==` and `inspect`) ,but the `#hash` logic isn't changed at all, and it doesn't have a dependency on any other methods in this class.

haberman commented:
@skippy Thanks -- sorry I wasn't meaning to suggest that this PR was causing the error (it couldn't possibly, since this PR hasn't been merged yet :), but wondering more whether this PR might actually fix the breakage, or something like that. I haven't come up to speed on the `#hash` vs. `#toHash` issue and didn't even realize they were separate, sorry! Thanks for the info.

I do want to get back to this PR to un-block you but I'm missing the bigger picture since I took over from Chris in the thick of things. Can you give me a big-picture overview of what this change is trying to accomplish?

Thanks,
Josh

haberman commented:
Oh scratch that actually, I took a closer look and I totally get it now. `#to_h` is for converting to a *hash table*, not to compute a hash of the container. You're just adding `#to_h` which didn't exist before. Ok got it now, sorry for the confusion. :)

skippy commented:
#### I updated the top level comment

Yeah, this PR is becoming a bit of a mess.  I found 4 non-related bugs
- **FIXED:** `repeated_field.rb` had an issue where the `#each_index` behavior is different when under an array or under an enumerable (this is probably an issue with ruby core).  The behavior was like it was as an enumerable, but since RepeatedArray is attempting to quack like an Array, I changed it to behave like `Array#each_index`

- **FIXED:** a bunch of tests and some of the code failed when run under jruby-9.0.0.0.pre2; these are fixed

- **NOT FIXED:** when an ProtoBuf object is converted to a JSON object, under JRuby, it returns the expected result.  However, under MRI, it does not return the default values.  See: https://github.com/google/protobuf/pull/396/files#diff-913cae3cb2a96306e220b73ff3d2cecbR118
- **NOT FIXED:** there are a bunch of new unit tests, including making sure encoding returns an expected result.  It actually returns a different value whether run under MRI or JRuby.  see: https://github.com/google/protobuf/pull/396/files#diff-913cae3cb2a96306e220b73ff3d2cecbR159


Besides those remaining 2 items (which should probably be separate tickets, and have the failing test commented out.. you can see that it failed in travis/jruby), I think this PR is good to go.  I would be more than happy to close this PR down and submit a clean one without the Jruby-9000 and repeated_field fixes and put those in separate PRs.

let me know if you would quickly like to chat offline and go through them in more detail

skippy commented:
@haberman thanks.  I missed your notes after I sent my update.  You are spot on... this is to add `Map#to_h`.

skippy commented:
@haberman any thoughts?  Let me know what you would like to have me do regarding the other possible bugs that were found (their is a broken test which highlights one of the bugs).  Then I'll update this PR and hopefully it should be good to go.

haberman commented:
Hi Adam,

Sorry again for the delay, and thanks for all the info.

I think that splitting this CL apart into smaller, more self-contained PRs would be very helpful.

When it comes to serializing defaults in JSON, both behaviors are technically correct! proto3 allows either, and some implementations let the user choose at runtime whether to serialize defaults to JSON or not.

I would be very interested to know exactly how the two serialized protos are differing.

ebenoist commented:
Is there any movement on getting this merged? This is still broken in the current beta releases.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok to test.

haberman commented:
Wow, this has been open for a while.

This PR is kind of big and addresses several related issues. It would be good to break it up into some more focused changes, and maybe even have issues filed for the specific problems that still remain.

ebenoist commented:
If @skippy cannot split this up, I'd be willing to take a look at getting just `to_hash` working. I was bit by the broken map support for the current Ruby codegen support in the latest beta.

skippy commented:
hey folks, my apologies for going silent on this pr.  I would love to revisit this PR but I'm pretty sure I won't be able to get to it for awhile yet.  But if I can help answer questions about what I was trying to do, please let me know!

dim commented:
Any chance for this PR to be revived? It's a really annoying bug. A :+1: from me

ebenoist commented:
Yeah, it's been a 3.0 blocker for me as well. I mentioned earlier that I'd be happy to split up @skippy's work if thats what the maintainers would like. I'm a little worried at the bit rot on this stuff at this point though.

haberman commented:
This PR is really out of date, and some of the issues have been fixed in other PRs, like this one: https://github.com/google/protobuf/pull/2847

I'm going to close this for now. Please feel free to re-open if you want to revive it.
